### ISSUE TITLE
nvidia-container-toolkit service generic-ressources not working

### ISSUE THREAD

**Issue by lyra00** [NVIDIA/nvidia-docker#1036]
_The template below is mostly useful for bug reports and support questions. Feel free to remove anything which doesn't apply to you and add more information where it makes sense._

_Also, before reporting a new issue, please make sure that:_

- _You read carefully the [documentation and frequently asked questions](https://github.com/NVIDIA/nvidia-docker/wiki)._
- _You [searched](https://github.com/NVIDIA/nvidia-docker/issues?utf8=%E2%9C%93&q=is%3Aissue) for a similar issue and this is not a duplicate of an existing one._
- _This issue is not related to [NGC](https://github.com/NVIDIA/nvidia-docker/wiki/NGC), otherwise, please use the [devtalk forums](https://devtalk.nvidia.com/default/board/200/nvidia-gpu-cloud-ngc-users/) instead._
- _You went through the [troubleshooting](https://github.com/NVIDIA/nvidia-docker/wiki/Troubleshooting) steps._

---

### 1. Issue or feature description
I'ved switched from NvidiaDocker2 to nvidia-container-toolkit.
When starting a container with 
docker run --gpus all nvidia/cuda nvidia-smi
everything working fine.

Starting this as a service with generic-resources it cant find nvidia-smi. The gpu is not bound to the Docker container.

My question is how to advertise the gpus to a service with nvidia-runtime-toolkit on the worker installed.
Is it still possible the old way througth generic-ressource.
Can someone Document it properly, because i cant find anything about it here.

### 2. Steps to reproduce the issue

1. Setup 2 Machines: one swarm manager, and one worker. The worker have nvidia GPUs with drivers/cuda installed.
2. Install docker-ce 19.03 on each machine and create a swarm.
**On the Worker**
3. Install nvidia-container-toolkit on the Worker.
4. Create a daemon.json file and add the following line:
{
    "node-generic-resources": ["NVIDIA-GPU=GPU-UUID"]
}
*You can find the GPU-UUID with nvidia-smi -a
5. heading to /etc/nvidia-container-runtime/config.toml and uncommenting/add following line swarm-resource = "DOCKER_RESOURCE_GPU"
6. add node.label.gpu = nvidia
7. reboot system
8. run **nvidia-smi** localy to ensure driver is working localy.
9. run **docker run --gpus all nvidia/cuda nvidia-smi** to ensure nvidia-docker runtime is working
10. create a service like that **docker service create --name smi  --constraint node.labels.gpu==nvidia --generic-resource "NVIDIA-GPU=1" nvidia/cuda nvidia-smi**

Step 8 and 9 are woking fine.
Step 10 failed with 

> starting container failed: OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"nvidia-smi\": executable file not found in $PATH": unknown

### 3. Information to [attach](https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/) (optional if deemed irrelevant)

 - [x] Some nvidia-container information: `nvidia-container-cli -k -d /dev/tty info`

[nvidia.txt](https://github.com/NVIDIA/nvidia-docker/files/3464052/nvidia.txt)

 - [x] Docker version from `docker version`

`Client: Docker Engine - Community
 Version:           19.03.1
 API version:       1.40
 Go version:        go1.12.5
 Git commit:        74b1e89
 Built:             Thu Jul 25 21:21:07 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.1
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.5
  Git commit:       74b1e89
  Built:            Thu Jul 25 21:19:36 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683`

 - [x] NVIDIA packages version from `dpkg -l '*nvidia*'` _or_ `rpm -qa '*nvidia*'`

`libnvidia-container-tools-1.0.3-1.x86_64
nvidia-driver-NVML-430.40-1.fc30.x86_64
nvidia-driver-libs-430.40-1.fc30.x86_64
akmod-nvidia-430.40-1.fc30.x86_64
nvidia-container-toolkit-1.0.1-2.x86_64
nvidia-persistenced-430.40-1.fc30.x86_64
nvidia-kmod-common-430.40-1.fc30.noarch
kmod-nvidia-5.1.20-300.fc30.x86_64-430.40-1.fc30.x86_64
nvidia-driver-cuda-libs-430.40-1.fc30.x86_64
nvidia-driver-cuda-430.40-1.fc30.x86_64
libnvidia-container1-1.0.3-1.x86_64`

 - [x] NVIDIA container library version from `nvidia-container-cli -V`

`version: 1.0.3
build date: 2019-08-02T23:16+0000
build revision: bfb23a2ed0c2e045f4586a367e989883f2752821
build compiler: gcc 4.8.5 20150623 (Red Hat 4.8.5-36)
build platform: x86_64
build flags: -D_GNU_SOURCE -D_FORTIFY_SOURCE=2 -DNDEBUG -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression -Wl,-zrelro -Wl,-znow -Wl,-zdefs -Wl,--gc-sections`

 - [x] NVIDIA container library logs (see [troubleshooting](https://github.com/NVIDIA/nvidia-docker/wiki/Troubleshooting))

no logs are being generated

 - [x] Docker command, image and tag used

docker run --gpus all nvidia/cuda nvidia-smi <- **Working**
docker service create --name smi  --constraint node.labels.gpu==nvidia --generic-resource "NVIDIA-GPU=1" nvidia/cuda nvidia-smi <- **Not Working**

**Comment by RenaudWasTaken** [comment]
Hello!

The `--gpus` option enabled by nvidia-container-toolkit doesn't work with generic resources.
You'll have to keep using the nvidia-docker2 package.

This is work in progress :)

**Comment by muru** [comment]
Is there a PR or issue we can subscribe to for this? There's something similar going on with Docker compose (https://github.com/docker/compose/issues/6691, https://github.com/docker/docker-py/pull/2419 via https://github.com/docker/docker-py/issues/2395), but I couldn't find anything for Docker Swarm.

**Comment by AndrewJDR** [comment]
@RenaudWasTaken It really would be best if there were a canonical issue to subscribe to so we can know when the latest nvidia-container-toolkit can be supported by docker compose / docker stack.

As @muru mentioned, there are several projects blocked on this, and it's causing a lot of confusion for people if you read those issues he linked to.
